### DISCUSSION TITLE
The adjoint method of IDAS in CasADi cannot generate correct derivatives for algebraic variables

### DISCUSSION THREAD

**Discussion by eaglema** [casadi/casadi#3097] [Q&A]
Hi all,

I found that **IDAS** in CasADi cannot generate the correct derivative for **algebraic variables,** i.e. $\frac{dz}{dp}$ when the **adjoint method** is used, although the forward method works fine for both state and algebraic variables, and the adjoint method can generate correct derivatives for state variables $\frac{dx}{dp}$. Please check the following demo code to see the problem.

import casadi as cas

#Simple DAE problem
x = cas.SX.sym('x')
z = cas.SX.sym('z')
p = cas.SX.sym('p')

ode = -p * x
alg = z - 2 * x

dae_prob = {'x': x, 'z': z, 'p': p, 'ode': ode, 'alg': alg}
dae_fun = cas.integrator('dae_fun', 'idas', dae_prob, 0, 1)

#Get dx/dp using forward method
dxdp_fwd = dae_fun.factory('dxdp_fwd', ['x0', 'p', 'fwd:p'], ['fwd:xf'])
res_dxdp_fwd = dxdp_fwd(x0=[1], p=[1], fwd_p=[1])

#Get dz/dp using forward method
dzdp_fwd = dae_fun.factory('dzdp_fwd', ['x0', 'p', 'fwd:p'], ['fwd:zf'])
res_dzdp_fwd = dzdp_fwd(x0=[1], p=[1], fwd_p=[1])

#Get dx/dp using adjoint method
dxdp_adj = dae_fun.factory('dxdp_adj', ['x0', 'p', 'adj:xf'], ['adj:p'])
res_dxdp_adj = dxdp_adj(x0=[1], p=[1], adj_xf=[1])

#Get dz/dp using adjoint method
dzdp_adj = dae_fun.factory('dzdp_adj', ['x0', 'p', 'adj:zf'], ['adj:p'])
res_dzdp_adj = dzdp_adj(x0=[1], p=[1], adj_zf=[1])

print('dx/dp when using forward method:', res_dxdp_fwd['fwd_xf'])
print('dx/dp when using adjoint method:', res_dxdp_adj['adj_p'])
print('dz/dp when using forward method:', res_dzdp_fwd['fwd_zf'])
print('dz/dp when using adjoint method:', res_dzdp_adj['adj_p'])

Here are the output of the code:
dx/dp when using forward method: -0.367879
dx/dp when using adjoint method: -0.367878
dz/dp when using forward method: -0.735759
dz/dp when using adjoint method: -1.54001e-34

I guess there is a bug in the CasADi IDAS interface. Please let me know if I'm wrong.

Best wishes,
Yingjie

**Comment by eaglema** [comment]
I found that the issue has been reported previously and will be resolved in 3.6.1. So I will close my discussion.

**Comment by jaeandersson** [comment]
Right, this is the corresponding issue: #2806
